core/src/main/java/org/elasticsearch/common/util/IndexFolderUpgrader.java

```diff
@@ -0,0 +1,265 @@
+                            final Path targetFile = targetDir.resolve(file.getFileName());
+                            /* We are pessimistic and do a copy first to the other path and then and atomic move
+                               to rename it such that in the worst case the file exists twice but is never lost
+                               or half written.
+                            */
+                            final Path targetTempFile = Files.createTempFile(targetDir,
+                                "upgrade_", "_" + file.getFileName().toString());
+                            Files.copy(file, targetTempFile, StandardCopyOption.COPY_ATTRIBUTES,
+                                StandardCopyOption.REPLACE_EXISTING);
+                            // we are on the same FS - this must work otherwise all bets are off
+                            Files.move(targetTempFile, targetFile, StandardCopyOption.ATOMIC_MOVE);
+                            Files.delete(file);
+                            movedFiles.add(targetFile);
+                        }
+                    }
+                }
+            }
+        }
+        if (movedFiles.isEmpty() == false) {
+            // fsync later it might be on disk already
+            logger.info("{} fsync files", targetPath.getShardId());
+            for (Path moved : movedFiles) {
+                logger.info("{} syncing [{}]", targetPath.getShardId(), moved.getFileName());
+                IOUtils.fsync(moved, false);
+            }
+            logger.info("{} syncing directory [{}]", targetPath.getShardId(), targetDir);
+            IOUtils.fsync(targetDir, true);
+        }
+    }
+
+    /**
+     * Runs check-index on the target shard and throws an exception if it failed
+     */
+    void checkShard(final ShardPath targetPath) throws IOException {
+        BytesStreamOutput os = new BytesStreamOutput();
+        PrintStream out = new PrintStream(os, false, StandardCharsets.UTF_8.name());
+        try (Directory directory = new SimpleFSDirectory(targetPath.resolveIndex());
+             final CheckIndex checkIndex = new CheckIndex(directory)) {
+            checkIndex.setInfoStream(out);
+            CheckIndex.Status status = checkIndex.checkIndex();
+            out.flush();
+            if (!status.clean) {
+                logger.warn("check index [failure]\n{}", new String(os.bytes().toBytes(), StandardCharsets.UTF_8));
+                throw new IllegalStateException("index check failure");
+            }
+        }
+    }
+
+    /**
+     * Upgrades all indices found under <code>nodeEnv</code>. Already upgraded indices are ignored.
+     */
+    public static void upgrade(final Settings settings, final NodeEnvironment nodeEnv) throws IOException {
+        XContentType format = XContentType.fromMediaTypeOrFormat(settings.get(MetaStateService.FORMAT_SETTING, "smile"));
```

Comment:
looking at the MetaDataStateFormat code, we should be able to read the format from the file - something stinks here :) . Also (different change) I think we should remove that setting and always write smile (and read what ever we find).